test: Streamline test suite a little

aequitas opened the pull request:
This change aims to streamline the integrationtest suite and reduce friction when running (parts of) test suites. 

Changes:
- add `test-testname` to makefile to easily run one suite
- made skipping tests first class citizen in test.sh (all available testsuites and the reasons they are skipped are now better exposed and discovered)
- suppress some output of gotest unless there is an error
- no longer build binary for examples and gotest suites
- removed .SILENT from makefile as it being applied to only some targets makes it feel weird (I just learned about this option btw, feel free to comment on this change)
- move individual tests out of `test.sh` and into `test-misc.sh`
- introduced the concept of testsuites to `test.sh`

Points for discussion:
- started effort to make scripts shellcheck compatible (might be to much unrelated changes in one PR, might take these out)

purpleidea commented on Makefile:
```diff
@@ -141,9 +140,15 @@ clean:
+test: build
 	./test.sh
 
+# create all targets for make tab completion
```
this is cool :)

aequitas replied:
thx, if for one like make, good choice for this project :)

purpleidea commented on Makefile:
```diff
@@ -141,9 +140,15 @@ clean:
 	./test.sh
 
+# create all targets for make tab completion
+test_suites=$(shell find test -name test-* -exec basename {} .sh \;)
```
shouldn't this be test/test-*.sh with a maxdepth? (so we don't get the subtests?)

aequitas replied:
true

purpleidea commented on docs/quick-start-guide.md:
```diff
@@ -99,6 +99,23 @@ At the moment we have:
 
 Please contribute more! We'd especially like to see a Debian package!
 
+### Testing
```
I don't think this should be part of the quick-start guide. It's already too big and is meant to be just the basics for new users to do "hello world".

aequitas replied:
i'll add a developer docs :)

purpleidea commented on test/util.sh:
```diff
@@ -1,15 +1,38 @@
+	"$@" || failures=$( [ -n "$failures" ] && echo "$failures\\n$@" || echo "$@" )
+}
+
+# colors https://gist.github.com/daytonn/8677243
```
typo: should be colours ;)

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,99 @@
+# runs all (or selected) test suite(s) in test/ and aggregates results
+# Usage:
+# 	./test.sh
+#   ./test.sh gofmt
```
In general, I've avoided places where I had more than one space, unless explicitly necessary. Please switch to tabs if you want indentation.

aequitas replied:
i'm a python dev so that will take some getting used to ;), but willdo :)

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,99 @@
-run-test diff <(tail -n +$start AUTHORS | sort) <(tail -n +$start AUTHORS)
+# run a test but expect it to fail (this will count as failure if the test passes!,
+# in that case it can be turned back to a normal run-test)
+# function xfail-test()
```
Is something missing here?

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,99 @@
 	exit 1
 fi
-echo 'ALL PASSED'
+green 'SUCCESS, ALL PASSED, AWESOME!'
```
your excitement is leaking :P

aequitas replied:
sorry :)

purpleidea commented on test/test-misc.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+# check that our examples still build, even if we don't run them here
```
what does this comment mean?

purpleidea commented on test/test-misc.sh:
```diff
@@ -0,0 +1,29 @@
+echo running "$0"
+
+ROOT=$(dirname "${BASH_SOURCE}")/..
+cd "${ROOT}" || exit
```
shouldn't this be exit 1? (other places too)

aequitas commented:
Impression of the colors:
![screen shot 2018-02-11 at 21 41 03](https://user-images.githubusercontent.com/365827/36078247-523c0cfc-0f74-11e8-9f95-6a94ce3fe5ca.png)

![screen shot 2018-02-11 at 21 41 58](https://user-images.githubusercontent.com/365827/36078259-68e97cdc-0f74-11e8-96e4-2526d81dcb82.png)

aequitas commented:
with regards to the colors, i will split it in a separate commit, also i think it should be possible to do interactive terminal detected and only enable colors for those (with an extra flag to disable them)

aequitas commented:
I removed all the shellcheck related changes and am working on full shellcheck compliance in a separate branch, PR will come later today.

Colo**u**r and test suite improvements have been separated and I added TERM detection and a disable flag.

Imho colo**u**rs can be on by default. I verified and travis uses TERM=xterm so there won't be any colours there. Also checked the Jenkins instance, it uses TERM=screen, also disabled colors for that.

purpleidea commented on Makefile:
```diff
@@ -17,7 +17,6 @@
 
 SHELL = /usr/bin/env bash
 .PHONY: all art cleanart version program lang path deps run race bindata generate build crossbuild clean test gofmt yamlfmt format docs rpmbuild mkdirs rpm srpm spec tar upload upload-sources upload-srpms upload-rpms copr
-.SILENT: clean bindata
```
why are we dropping ths?

aequitas replied:
> removed .SILENT from makefile as it being applied to only some targets makes it feel weird (I just learned about this option btw, feel free to comment on this change)

Adding `lang` to it instead of removing is also fine imho, but i was confused at first why `bindata` and `lang` had different behaviour while the recipies where the same.

purpleidea replied:
So what should we do here?
The reason (IIRC) these are here is because they're uninteresting, whereas getting output from the `lang` target is important.

aequitas replied:
Lets keep it as is.

purpleidea commented on Makefile:
```diff
@@ -141,9 +140,15 @@ clean:
 	rm -f *_mock.go		# generated by `go generate`
 
-test: bindata
+test: build
```
good idea :)

purpleidea commented on docs/development.md:
```diff
@@ -0,0 +1,21 @@
+# Development
+Additional information and help regarding developing `mgmt`. Useful tools, conventions, etc.
```
Maybe this should be a sentence instead? Like "This contains some additional..."

aequitas replied:
Correct.

purpleidea commented on docs/development.md:
```diff
@@ -0,0 +1,21 @@
+
+Be sure to read [quick start guide](docs/quick-start-guide.md) first.
+
+## Testing
```
This is cool to have in the docs, but please add some sort of line early on that says something along the lines of:

"Native golang tests are preferred over tests written in our shell testing framework. Please see https://golang.org/pkg/testing/ for more information."

aequitas replied:
Will do

aequitas replied:
Adding 'test-suites' is actually a bad pattern, shell tests should be preferred over testsuites and go tests should be preferred over them. I don't know what I was thinking when I wrote this. I'll adjust it to focus on gotest and shell test.

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,92 @@
+	# show skip message only when running full suite
+	if test -z "$testsuite";then
+		echo skipping "$@" "($REASON)"
+		blue SKIP
```
colours don't exist in this commit yet.

aequitas replied:
I've missed some when splitting up the commits, i'll correct this and the others

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,92 @@
+run-testsuite ./test/test-headerfmt.sh
+run-testsuite ./test/test-commit-message.sh
+run-testsuite ./test/test-govet.sh
+run-testsuite ./test/test-gotest.sh
```
I'm okay that you changed the order on these last two, it probably makes more sense, but was it a conscious decision, and if so, why.

aequitas replied:
Good you question this! gofmt was intentional because it is quick and might give you something to work on already while waiting for the rest of the suite to complete, gotest/examples was not intentional, i'll reverse them back

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,92 @@
 	exit 1
 fi
-echo 'ALL PASSED'
+echo 'SUCCESS, ALL PASSED, AWESOME!'
```
Sorry, nack. You get to make these sorts of changes once you've got more golang code committed first ;)

purpleidea commented on test.sh:
```diff
@@ -1,51 +1,92 @@
 if [[ -n "$failures" ]]; then
-	echo 'FAIL'
+	echo
+	echo 'FAILURE'
```
Please don't rename this line unless there's a good reason.

aequitas replied:
The reason behind it is to distinct the conclusion of the entire run from the output of a single test, PASS/FAIL SUCCESS/FAILURE. Any other suggestion is welcome, if you don't like it I'm just as happy to keep it as FAIL.

purpleidea replied:
Let's keep this kind of churn to a minimum for now please. Sorry.

purpleidea commented on test/test-gotest.sh:
```diff
@@ -7,29 +7,31 @@ ROOT=$(dirname "${BASH_SOURCE}")/..
 base=$(go list .)
 for pkg in `go list ./... | grep -v "^${base}/vendor/" | grep -v "^${base}/examples/" | grep -v "^${base}/test/" | grep -v "^${base}/old/" | grep -v "^${base}/tmp/"`; do
-	echo "Testing: $pkg"
+	echo "  testing: $pkg"
```
Nack. I especially dislike indentation by spaces, sorry.

aequitas replied:
Sorry about the spaces, it's hard to get out of my system. The reason behind this was that this is the only test showing progress output during its run and I wanted to offset that from the 'suite' output.

purpleidea replied:
You are welcome to use as many tabs as you like! :)

purpleidea commented on test/test-misc.sh:
```diff
@@ -0,0 +1,29 @@
+run-test diff <(tail -n +$start AUTHORS | sort) <(tail -n +$start AUTHORS)
+
+if [[ -n "$failures" ]]; then
+	redb 'FAIL'
```
colours don't exist in this commit.

purpleidea commented on test/test-misc.sh:
```diff
@@ -0,0 +1,29 @@
+	echo
+	exit 1
+fi
+greenb 'PASS'
```
colours...

aequitas commented:
I addressed the comments and also added individual test target for shell tests (eg: `make test-shell-load0`).

purpleidea commented on Makefile:
```diff
@@ -17,11 +17,9 @@
 
 # a large amount of output from this `find`, can cause `make` to be much slower!
 GO_FILES := $(shell find * -name '*.go' -not -path 'old/*' -not -path 'tmp/*')
-
```
leave this random nl please

purpleidea commented on Makefile:
```diff
@@ -157,9 +155,20 @@ clean:
 	./test.sh
 
+# create all test targets for make tab completion (eg: make test-gofmt)
+test_suites=$(shell find test -maxdepth 1 -name test-* -exec basename {} .sh \;)
```
is it more appropriate to use `find test/` (with trailing slash so it's more readable and explicit)
(same below)

aequitas replied:
I agree, that was my first approach, but then you end up with double slashes in paths like:
```
~/.g/s/g/p/mgmt (testscripting|REBASE-i 1/2|✚17…2) $ find test/
test/
test//omv
test//omv/helloworld.yaml
test//omv/pkg1a.yaml
test//omv/pkg1b.yaml
```

Which technically doesn't matter at all i think, but causes me to go into correctness limbo :).

But i think we can favor for clearness in the Makefile.

purpleidea commented on docs/development.md:
```diff
@@ -0,0 +1,31 @@
+Be sure to read [quick start guide](docs/quick-start-guide.md) first.
+
+## Testing
+This project has both unit tests in the form of Go tests and integration tests using shell scripting.
```
s/Go/golang/ for unambiguous-ness please.

purpleidea commented on docs/development.md:
```diff
@@ -0,0 +1,31 @@
+## Testing
+This project has both unit tests in the form of Go tests and integration tests using shell scripting.
+
+Native Golang tests are preferred over tests written in our shell testing framework. Please see https://golang.org/pkg/testing/ for more information.
```
s/Golang/golang/

purpleidea commented on docs/development.md:
````diff
@@ -0,0 +1,31 @@
+make test
+```
+
+There is a library of quick and small integration tests for Language or YAML related things, check out `test/shell/`. Adding a test is as easy as copying one of the files in `test/shell/` and adapting it.
````
These can be relative links with `[blah](url)`

s/Language or/the language and/

purpleidea commented on test.sh:
```diff
@@ -1,54 +1,95 @@
 	exit 1
 fi
-echo 'ALL PASSED'
+echo 'SUCCESS'
```
Unnecessary churn. Nack for now please.

purpleidea commented on test/test-misc.sh:
```diff
@@ -0,0 +1,29 @@
+echo running "$0"
+
+ROOT=$(dirname "${BASH_SOURCE}")/..
+cd "${ROOT}" || exit 1
```
if you're going to add this `|| exit 1` on here, please do it to the other `cd "${ROOT}"` 's too.

aequitas replied:
That will be part of the Shellcheck commit which is not included in this PR. For new/changed lines I already made them shellcheck compatible, for unchanged lines I left hem in that commit for separation.

purpleidea commented on test/util.sh:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env bash
```
nack,

`#!/bin/bash`

please

richarson replied:
For the record, I believe FreeBSD has bash in /usr/bin, if you're ever going to support the BSDs.
Sorry for hijacking :)

aequitas replied:
@richarson I discussed it with @purpleidea in irc, I had no real argument for it, only that is was the default template in my text editor :). I believe there was no major argument against it as well. But I think having consistent conventions in the codebase is more valuable than my template's preference ;)

I'm working on a Shellcheck compliance commit, we might discus this later on there if needed or wait until someone complains about it.

purpleidea commented:
Your tab completion in Make to run tests is *SO UNBELIEVABLY DOPE* and a "great idea"(tm) that it is merged. I don't even care about the `;then` stuff. It's in.

purpleidea commented:
Sorry on the delay here btw, travis is experiencing issues https://www.traviscistatus.com/ so I'm waiting on that...

aequitas commented:
Somehow I dropped the colours commit from a previous push. Let me know what you want with it.

Current branch is rebased on latest master

aequitas commented:
I added a commit that fixed the golint test as it now needs to exclude generated files since: https://github.com/purpleidea/mgmt/commit/b34b35986020621b298a1b8fa12f29cada4de830

purpleidea commented:
https://github.com/purpleidea/mgmt/pull/329/commits/76bb52f1144c25c63636b857877910e0bb0f707b is merged :)

aequitas commented on test/test-golint.sh:
```diff
@@ -26,7 +26,9 @@ if [ "$COMMITS" != "" ] && [ "$COMMITS" -gt "1" ]; then
 
-LINT=`find . -maxdepth 3 -iname '*.go' -not -path './old/*' -not -path './tmp/*' -not -path './bindata/*' -exec golint {} \;`	# current golint output
+# find all go files, exluding temporary directories and generated files
+LINT=$(find * -maxdepth 2 -iname '*.go' -not -path './old/*' -not -path './tmp/*' -not -path './bindata/*' -not -path 'lang/y.go' -not -path 'lang/lexer.nn.go' -exec golint {} \;)	# current golint output
```
I changed '.' to '*' for find. But I think instead I had to add the 'lang/...' excludes as './lang/...'

purpleidea commented:
I'm going to NACK the colours commit for now (sorry) but keep it on my "think about it" list. I've pushed that commit to feat/colours on my branch.

purpleidea commented:
@aequitas I've merged a bunch of patches in this PR, but it's gotten unwieldy, so I am going to close it.

Please submit anything outstanding in separate PR's. Thank you again!

aequitas commented:
@purpleidea sure thing, I'll see whats left open and create a new PR where needed.
